File: assets/static/blog_posts/algo/datastructures/stacks/implementation.ex


defmodule Stack do

  defstruct [data: [], size: 0]

  def new do; %Stack{} end

  def push(stack, item) do
    %Stack{
      data: [item | stack.data],
      size: stack.size + 1
    }
  end

  def pop(%Stack{ data: [] }) do; {:empty, new()} end
  def pop(%Stack{ data: [head] }) do; {head, new()} end
  def pop(%Stack{ data: [head | tail], size: size }) do
    {head, %Stack{data: tail, size: size-1}}
  end

  def peek(%Stack{}) do; :empty end
  def peek(%Stack{data: [head | _tail]}) do; head end
end

File: assets/static/blog_posts/algo/datastructures/stacks/implementation.py
class Stack:
    def __init__(self):
        self.data = []

    def __repr__(self):
        return str(self.data)

    def push(self, item):
        self.data.append(item)

    def pop(self):
        val = self.data[-1]
        self.data = self.data[:-1]
        return val

    def peek(self):
        return self.data[-1]
